src/util/simplify_expr_array.cpp

```diff
@@ -31,86 +36,77 @@ bool simplify_exprt::simplify_index(exprt &expr)
        index.op0().op1()==index.op1())
     {
       exprt tmp=index.op0().op0();
-      expr.op1()=tmp;
+      index = tmp;
       no_change = false;
     }
     else if(index.op0().id()==ID_mult &&
             index.op0().operands().size()==2 &&
             index.op0().op0()==index.op1())
     {
       exprt tmp=index.op0().op1();
-      expr.op1()=tmp;
+      index = tmp;
       no_change = false;
     }
   }
 
-  if(array.id()==ID_lambda)
+  if(array.id() == ID_lambda)
   {
     // simplify (lambda i: e)(x) to e[i/x]
 
     const auto &comprehension = to_array_comprehension_expr(array);
 
-    if(expr.op1().type() == comprehension.arg().type())
+    if(index.type() == comprehension.arg().type())
     {
       exprt tmp = comprehension.body();
-      replace_expr(comprehension.arg(), expr.op1(), tmp);
-      expr.swap(tmp);
-      simplify_rec(expr);
-      return false;
+      replace_expr(comprehension.arg(), index, tmp);
+      simplify_rec(tmp);
+      return std::move(tmp);
     }
   }
   else if(array.id()==ID_with)
   {
     // we have (a WITH [i:=e])[j]
 
     if(array.operands().size() != 3)
-      return true;
+      return unchanged(expr);
 
     const auto &with_expr = to_with_expr(array);
 
-    if(with_expr.where() == expr.op1())
+    if(with_expr.where() == index)
     {
       // simplify (e with [i:=v])[i] to v
-      exprt tmp = with_expr.new_value();
-      expr.swap(tmp);
-      return false;
+      return with_expr.new_value();
     }
     else
     {
       // Turn (a with i:=x)[j] into (i==j)?x:a[j].
       // watch out that the type of i and j might be different.
       const exprt rhs_casted =
-        typecast_exprt::conditional_cast(with_expr.where(), expr.op1().type());
-
-      equal_exprt equality_expr(expr.op1(), rhs_casted);
-
-      simplify_inequality(equality_expr);
+        typecast_exprt::conditional_cast(with_expr.where(), index.type());
 
-      index_exprt new_index_expr(with_expr.old(), expr.op1(), expr.type());
+      exprt equality_expr = simplify_inequality(equal_exprt(index, rhs_casted));
 
-      simplify_index(new_index_expr); // recursive call
+      exprt new_index_expr = simplify_index(index_exprt(
+        with_expr.old(), index, index_expr.type())); // recursive call
 
       if(equality_expr.is_true())
       {
-        expr = with_expr.new_value();
-        return false;
+        return with_expr.new_value();
       }
       else if(equality_expr.is_false())
       {
-        expr.swap(new_index_expr);
-        return false;
+        return new_index_expr;
       }
 
       if_exprt if_expr(equality_expr, with_expr.new_value(), new_index_expr);
-      expr = simplify_if(if_expr).expr;
-      return false;
+      return simplify_if(if_expr).expr;
```

Comment:
It might not make much of a difference, but elsewhere we seem to use `return changed(simplify_if(if_expr));`

Reply:
ok